File: src/Chapter4/SoldComponent.java
package Chapter4;

/*                                                                            *
 * @author Gianluca Capone & Davide Sgobba                                    *
 *                                                                            */

/* This class contains all variables that define the component product        *
 * produced by specialized component firms and the methods that operate over  *
 * these variables                                                            */
class SoldComponent {
    
    // VARIABLES
    double mod;
    /* Merit of design of the component product (M-CO_f,t)                    */
    double muProg;
    /* Mean of the distribution of component technical progress (mu-CO_f,t)   */
    double productionCost;
    /* Component production cost (C-CO_f,t)                                   */
    double u;
    /* Propensity of component to be sold to computer firms (u_f,h,t)         */
    double U;
    /* Probability of component to be sold to computer firms (U_f,h,t)        */
    double uExt;
    /* Propensity of component to be sold to external users (u_f,h,t)         *
     * See footnone 11 for the differences between u/U and uExt/UExt          */
    double UExt;
    /* Probability of component to be sold to external users (U_f,h,t)        */
    
    //TECHNICAL VARIABLES & OBJECTS
    ComponentFirm Firm;
    /* Access to the firm                                                     */
    
    // CONSTRUCTOR
    SoldComponent(double MOD, ComponentFirm FIRM) {
        muProg         = 0;
        productionCost = 0;
        u              = 0;
        U              = 0;
        uExt           = 0;
        UExt           = 0;
        mod            = MOD;
        Firm           = FIRM;
    }

    /* This method computes the propensity to sell to computer firms          */
    void calcPropToSell() {
        /* Equation 2 adapted to components                                   */
        u = Math.pow(mod, Firm.CmpMarket.deltaMod) 
             * Math.pow(1 + Firm.share, Firm.CmpMarket.deltaShare[Firm.tId]);
    }
    
    /* This method computes the probability to sell to computer firms         */
    void calcProbToSell(double SUMRAT) {
        /* Equation 4 adapted to components                                   */
        U = u/SUMRAT;
    }
    
    /* This method computes the mod of the component based on the level of    *
     * past mod and the level of public knowledge                             */
    void calcMod() {
        
        double zmax = 0;
        /* Equation 14.b                                                      */
        muProg = Math.log(Firm.CmpMarket.pk[Firm.tId]) * (1 - Firm.CmpMarket.internalCum)
                             + Math.log(mod) * Firm.CmpMarket.internalCum;
        for (int i = 1; i <= Firm.numOfDrawsCmp; i++) {
            double z = Math.exp(muProg + Math.sqrt(Firm.CmpMarket.sdCmp[Firm.tId])
                       * Firm.CmpMarket.rng.nextGaussian());
            if (z > zmax) {
                zmax = z;
            }
        }
        if (zmax > mod) {
            mod = zmax;
        }
        if (mod > 0) {
            /* Equation 11                                                    */
            productionCost = Firm.CmpMarket.nu / mod;
        }
    }
    
    /* This method computes the propensity to sell to external markets        */
    void calcPropToSellExt() {
        /* Equation 2 adapted to components                                   */
        uExt = Math.pow(mod, Firm.CmpMarket.deltaMod) 
             * Math.pow(1 + Firm.share, Firm.CmpMarket.deltaShare[Firm.tId]);
    }
    
    /* This method computes the probability to sell to external markets       */
    void calcProbToSellExt(double SUMPTS) {
        /* Equation 4 adapted to components                                   */
        UExt = uExt/SUMPTS;
    }
    
    /* This method is activated when exit occurs: it resets the most relevant *
     * variables at the product level                                         */
    void exitComponent() {
        muProg         = 0;
        productionCost = 0;
        u              = 0;
        U              = 0;
        uExt           = 0;
        UExt           = 0;
        mod            = 0;
    }
    
}

File: src_py/Chapter4/sold_component.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
SoldComponent模块 - SoldComponent类的Python实现
转换自Java版本的SoldComponent.java
"""

import math

"""
@author Gianluca Capone & Davide Sgobba
Python转换

此类包含定义专业组件公司生产的组件产品的所有变量以及操作这些变量的方法
"""
class SoldComponent:
    
    def __init__(self, mod, firm):
        """
        构造函数
        
        Args:
            mod: 组件产品的设计优点 (M-CO_f,t)
            firm: 公司对象引用
        """
        # 变量
        self.mu_prog = 0.0         # 组件技术进步分布的均值 (mu-CO_f,t)
        self.production_cost = 0.0  # 组件生产成本 (C-CO_f,t)
        self.u = 0.0               # 组件销售给计算机公司的倾向 (u_f,h,t)
        self.U = 0.0               # 组件销售给计算机公司的概率 (U_f,h,t)
        self.u_ext = 0.0           # 组件销售给外部用户的倾向 (u_f,h,t)，见脚注11的u/U和uExt/UExt的区别
        self.U_ext = 0.0           # 组件销售给外部用户的概率 (U_f,h,t)
        self.mod = mod             # 组件的设计优点 (M-CO_f,t)
        
        # 技术变量和对象
        self.firm = firm           # 访问公司对象 
    
    def calc_prop_to_sell(self):
        """
        计算向计算机公司销售的倾向
        """
        # 方程2（适用于组件）
        self.u = (self.mod ** self.firm.cmp_market.delta_mod) * \
                ((1 + self.firm.share) ** self.firm.cmp_market.delta_share[self.firm.t_id])
    
    def calc_prob_to_sell(self, sum_rat):
        """
        计算向计算机公司销售的概率
        
        Args:
            sum_rat: 所有公司倾向的总和
        """
        # 方程4（适用于组件）
        self.U = self.u / sum_rat
    
    def calc_mod(self):
        """
        根据过去的mod水平和公共知识水平计算组件的mod
        """
        z_max = 0.0
        # 方程14.b
        self.mu_prog = ((1 - self.firm.cmp_market.internal_cum) * 
                         math.log(self.firm.cmp_market.pk[self.firm.t_id]) + 
                         self.firm.cmp_market.internal_cum * math.log(self.mod))
        
        for i in range(1, self.firm.num_of_draws_cmp + 1):
            z = math.exp(self.mu_prog + 
                       math.sqrt(self.firm.cmp_market.sd_cmp[self.firm.t_id]) * 
                       self.firm.cmp_market.rng.gauss(0, 1))
            if z > z_max:
                z_max = z
        
        if z_max > self.mod:
            self.mod = z_max
        
        if self.mod > 0:
            # 方程11
            self.production_cost = self.firm.cmp_market.nu / self.mod
    
    def calc_prop_to_sell_ext(self):
        """
        计算向外部市场销售的倾向
        """
        # 方程2（适用于组件）
        self.u_ext = (self.mod ** self.firm.cmp_market.delta_mod) * \
                    ((1 + self.firm.share) ** self.firm.cmp_market.delta_share[self.firm.t_id])
    
    def calc_prob_to_sell_ext(self, sum_pts):
        """
        计算向外部市场销售的概率
        
        Args:
            sum_pts: 所有公司倾向的总和
        """
        # 方程4（适用于组件）
        self.U_ext = self.u_ext / sum_pts
    
    def exit_component(self):
        """
        退出时激活的方法：重置产品层面的最相关变量
        """
        self.mu_prog = 0.0
        self.production_cost = 0.0
        self.u = 0.0
        self.U = 0.0
        self.u_ext = 0.0
        self.U_ext = 0.0
        self.mod = 0.0 